src/SourceFile.js

```diff
@@ -119,91 +148,82 @@ class SourceFile {
      * The content of the file, stored as regular Javascript string
      * encoded in UTF-8.
      *
-     * @type {String}
+     * @type {String|undefined}
      * @protected
      */
     content;
-    
+
     /**
      * Get the content of this file encoded as a regular Javascript
      * string.
-     * @returns {String} the content of the file, encoded as a JS string
+     * @returns {String|undefined} the content of the file, encoded as a JS string
```

Comment:
If we want to explicitly allow undefined content, I think its meaning should be explicitly defined like: "undefined means that the file has not yet been loaded from disk, so you need to run file.read() first or provide the content yourself if you're creating the file from scratch".

Personally I'm not sure if I like the idea of undefined content exposed in the interface, because it sort of forces you to acknowledge the possiblility of the file not being loaded every time you want to use it. One of the patterns I recall from C# is throwing `InvalidOperationException`, another is having a private constructor and static methods `SourceFile.read(path)` and `SourceFile.create(path, content)` which would ensure that `SourceFile` is always initialized with necessary data

Reply:
Well, what I was thinking is that if you are reading a file from disk, then content will be set to something. But, if you are creating a new file, then content can be undefined at first until you set the content and write it to disk. But, then we change SourceFile to be immutable, so now we can get rid of this again.